### PR TITLE
BlockCollector detects repeated block

### PR DESCRIPTION
**Pull request by hellosmallstone**
BlockCollector detects repeated block from same observer and exclude it from block counting

https://github.com/Hyperledger-TWGC/tape/issues/165

Signed-off-by: wuxu <wuxu1103@163.com>

### CODE REVIEW

File: pkg/infra/block_collector.go [hunk 1]
```diff
@@ -18,54 +18,68 @@ type BlockCollector struct {
 	thresholdP, totalP int
 	totalTx            int
 	registry           map[uint64]int
+	addressBlock       map[uint64]map[string]bool
+}
+
+// BlockAddress describe the source of block
+type BlockAddress struct {
```
Comment (guoger) [inline pkg/infra/block_collector.go:25]: `AddressedBlock` maybe?

and if we declare it with embedding:
```go
type AddressedBlock struct {
  *peer.FilteredBlock

  Address string
}
```
we could avoid some text changes in other part of code?

File: pkg/infra/block_collector.go [hunk 2]
```diff
@@ -18,54 +18,68 @@ type BlockCollector struct {
 	thresholdP, totalP int
 	totalTx            int
 	registry           map[uint64]int
+	addressBlock       map[uint64]map[string]bool
```
Comment (guoger) [inline pkg/infra/block_collector.go:21]: i suppose only one of these two maps is sufficient?

also, with 16 peers and several millions of blocks, this map could take significant memory space. I think we do need gc here.

File: pkg/infra/benchmark_test.go [hunk 3]
```diff
@@ -88,21 +89,24 @@ func BenchmarkSyncCollector16(b *testing.B) { benchmarkSyncCollector(16, b) }
...
+	blockAddr := make(chan *BlockAddress, 100)
 	done := make(chan struct{})
-	go instance.Start(context.Background(), block, done, b.N, time.Now(), false)
+	go instance.Start(context.Background(), blockAddr, done, b.N, time.Now(), false)
 
 	b.ReportAllocs()
 	b.ResetTimer()
 	for i := 0; i < concurrent; i++ {
-		go func() {
+		go func(num int) {
```
Comment (guoger) [inline pkg/infra/benchmark_test.go:99]: num -> idx

**Review by guoger — CHANGES_REQUESTED** [review]
I'm generally Okay with the approach, but with some minor concerns

File: pkg/infra/block_collector_test.go [hunk 4]
```diff
@@ -55,57 +56,57 @@ var _ = Describe("BlockCollector", func() {
 			instance, err := infra.NewBlockCollector(4, 4)
 			Expect(err).NotTo(HaveOccurred())
 
-			block := make(chan *peer.FilteredBlock)
+			block := make(chan *infra.AddressedBlock)
 			done := make(chan struct{})
 			go instance.Start(context.Background(), block, done, 2, time.Now(), false)
 
-			block <- &peer.FilteredBlock{Number: 1, FilteredTransactions: make([]*peer.FilteredTransaction, 1)}
+			block <- &infra.AddressedBlock{Address: "peer1", FilteredBlock: &peer.FilteredBlock{Number: 1, FilteredTransactions: make([]*peer.FilteredTransaction, 1)}}
```
Comment (SamYuan1990) [inline pkg/infra/block_collector_test.go:63]: em...提一个公共函数？

File: pkg/infra/block_collector_test.go [hunk 5]
```diff
@@ -119,17 +120,18 @@ var _ = Describe("BlockCollector", func() {
...
 
 			var wg sync.WaitGroup
 			wg.Add(100)
 			for i := 0; i < 100; i++ {
-				go func() {
+				go func(idx int) {
 					defer wg.Done()
-					block <- &peer.FilteredBlock{FilteredTransactions: make([]*peer.FilteredTransaction, 1)}
-				}()
+					block <- &infra.AddressedBlock{Address: fmt.Sprintf("peer%d", idx),
```
Comment (SamYuan1990) [inline pkg/infra/block_collector_test.go:132]: int转str就够了？

File: pkg/infra/block_collector.go [hunk 6]
```diff
@@ -10,19 +10,81 @@ import (
 	"github.com/pkg/errors"
 )
 
+type BitMap struct {
```
Comment (guoger) [inline pkg/infra/block_collector.go:13]: `BitMap` could be moved to its own file/pkg

File: pkg/infra/observer.go [hunk 7]
```diff
@@ -14,24 +14,26 @@ type Observers struct {
 }
 
 type Observer struct {
+	num     int
```
Comment (guoger) [inline pkg/infra/observer.go:17]: num -> index

File: pkg/infra/block_collector.go [hunk 8]
```diff
@@ -63,9 +127,6 @@ func (bc *BlockCollector) Start(
 			if cnt == bc.totalP {
 				// committed on all peers, remove from registry
 				delete(bc.registry, block.Number)
```
Comment (guoger) [inline pkg/infra/block_collector.go:129]: 如果我们从registry中删除区块，那么如果下次收到重复的块，如果判断呢？

**Review by guoger — CHANGES_REQUESTED** [review]

File: pkg/infra/block_collector.go [hunk 9]
```diff
@@ -91,11 +151,31 @@ func (bc *BlockCollector) Commit(block uint64) (committed bool) {
...
+
+// IsRepeatedBlock Determine whether the block from the address has been received
+// if not received, add it to the map and return false, and the number of received peers
+// otherwise, return true and zero
+func (bc *BlockCollector) IsRepeatedBlock(block *AddressedBlock) (bool, int) {
+	bitMap, ok := bc.registry[block.Number]
+	if !ok {
+		b, err := NewBitMap(bc.totalP)
+		if err != nil {
+			panic("Can not make new bitmap for BlockCollector" + err.Error())
```
Comment (SamYuan1990) [inline pkg/infra/block_collector.go:168]: no panic

File: pkg/infra/block_collector.go [hunk 10]
```diff
@@ -10,19 +10,81 @@ import (
...
 )
 
+type BitMap struct {
+	count      int // number of bits set
+	capability int // total number of bits
+	bits       []uint64
+}
+
+// Has determine whether the specified position is set
+func (b *BitMap) Has(num int) bool {
```
Comment (SamYuan1990) [inline pkg/infra/block_collector.go:20]: bitmap的话会不会泄露？如果peer足够多？

File: pkg/infra/observer.go [hunk 11]
```diff
@@ -60,8 +62,9 @@ func CreateObserver(ctx context.Context, channel string, node Node, crypto *Cryp
 	return &Observer{Address: node.Addr, d: deliverer, logger: logger}, nil
 }
 
-func (o *Observer) Start(errorCh chan error, blockCh chan<- *peer.FilteredBlock, now time.Time) {
-	o.logger.Debugf("start observer for orderer %s", o.Address)
+func (o *Observer) Start(errorCh chan error, blockCh chan<- *AddressedBlock, now time.Time) {
+	o.logger.Debugf("start observer for peer %s", o.Address)
+	n := 0
```
Comment (SamYuan1990) [inline pkg/infra/observer.go:67]: remove n here?

**Review by SamYuan1990 — CHANGES_REQUESTED** [review]

File: pkg/infra/block_collector.go [hunk 12]
```diff
@@ -35,70 +45,59 @@ func NewBlockCollector(threshold int, total int) (*BlockCollector, error) {
...
-//
-// Commit commits a block to collector. It returns true iff the number of peers on which
-// this block has been committed has satisfied thresholdP.
-func (bc *BlockCollector) Commit(block *peer.DeliverResponse_FilteredBlock, finishCh chan struct{}, now time.Time) (committed bool) {
-	bc.Lock()
-	defer bc.Unlock()
+// commit commits a block to collector.
+// If the number of peers on which this block has been committed has satisfied thresholdP,
+// adds the number to the totalTx.
+func (bc *BlockCollector) commit(block *AddressedBlock, finishCh chan struct{}, totalTx int, now time.Time, printResult bool) {
```
Comment (SamYuan1990) [inline pkg/infra/block_collector.go:67]: 加个 to do？

我在思考这里我们在这个函数里包括了。

构造其他对象。。。
方法。。。
常量对比。。。
提前退出。。。

我没想好怎么弄，但是感觉味道比较杂。

**Review by SamYuan1990 — APPROVED** [review]

File: pkg/infra/block_collector.go [hunk 13]
```diff
@@ -17,12 +18,21 @@ type BlockCollector struct {
...
 	thresholdP, totalP int
 	totalTx            int
-	registry           map[uint64]int
+	registry           map[uint64]*bitmap.BitMap
+}
+
+// AddressedBlock describe the source of block
+type AddressedBlock struct {
+	*peer.FilteredBlock
+	Address int // source peer's number
```
Comment (guoger) [inline pkg/infra/block_collector.go:27]: it would be nice to rename this field to `PeerIdx`

**Review by guoger — APPROVED** [review]

**Pull request merged**